devito/finite_differences/derivative.py

```diff
@@ -0,0 +1,140 @@
+import sympy
+
+from cached_property import cached_property
+
+from devito.tools import flatten
+from devito.finite_differences.finite_difference import (left, right, centered,
+                                                         generic_derivative, transpose,
+                                                         first_derivative, direct,
+                                                         cross_derivative)
+from devito.finite_differences.differentiable import Differentiable
+from devito.tools import as_tuple
+from devito.types import Dimension
+
+
+class Derivative(sympy.Derivative, Differentiable):
+
+    """
+    An unevaluated Derivative, which carries metadata (Dimensions,
+    derivative order, etc) to be evaluated.
+    """
+
+    def __new__(cls, expr, *dims, **kwargs):
+        print(expr, dims, kwargs)
+        # Verifies that there is one order per dimension if
+        # deriv_order is provided
+        deriv_order = kwargs.get('deriv_order', None)
+        ndim = len(as_tuple(dims))
+        norder = len(as_tuple(deriv_order)) if deriv_order is not None else 0
+
+        if deriv_order is not None and ndim != norder:
+            raise ValueError("Different number of Dimensions and derivative orders")
+    
+        if deriv_order is not None:
+            orders = dict((d, o) for d, o in zip(dims, as_tuple(deriv_order)))
+        else:
+            orders = dict()
+            for i, d in enumerate(dims):
+                if isinstance(d, (tuple, sympy.Tuple)):
+                    orders[d[0]] = d[1]
```

Comment:
However, I'm really confused... this is assuming a 2-element length tuple? 

I think we need lots of examples in the docstring, because this is more than impenetrable